// graphqlgen/dist/src/generated/graphqlgen.js
"use strict";
// Code generated by github.com/prisma/graphqlgen, DO NOT EDIT.
exports.__esModule = true;
var QueryResolvers;
(function (QueryResolvers) {
    QueryResolvers.defaultResolvers = {};
})(QueryResolvers = exports.QueryResolvers || (exports.QueryResolvers = {}));
var PostResolvers;
(function (PostResolvers) {
    PostResolvers.defaultResolvers = {
        id: function (parent) { return parent.id; },
        title: function (parent) { return parent.title; },
        content: function (parent) { return parent.content; },
        published: function (parent) { return parent.published; }
    };
})(PostResolvers = exports.PostResolvers || (exports.PostResolvers = {}));
var UserResolvers;
(function (UserResolvers) {
    UserResolvers.defaultResolvers = {
        id: function (parent) { return parent.id; },
        name: function (parent) { return (parent.name === undefined ? null : parent.name); }
    };
})(UserResolvers = exports.UserResolvers || (exports.UserResolvers = {}));
var MutationResolvers;
(function (MutationResolvers) {
    MutationResolvers.defaultResolvers = {};
})(MutationResolvers = exports.MutationResolvers || (exports.MutationResolvers = {}));
