// graphqlgen/dist/graphqlgen/src/tests/scaffold-typescript/basic.test.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var _this = this;
Object.defineProperty(exports, "__esModule", { value: true });
var path_1 = require("path");
var index_1 = require("../../index");
var validation_1 = require("../../validation");
var parse_1 = require("../../parse");
var relative = function (p) { return path_1.join(__dirname, p); };
test('basic schema', function () { return __awaiter(_this, void 0, void 0, function () {
    var _a, language, config, schema, modelMap, _b, generatedTypes, generatedResolvers;
    return __generator(this, function (_c) {
        language = 'typescript';
        config = (_a = {
                language: language,
                schema: relative('../fixtures/basic/schema.graphql'),
                context: relative('../fixtures/basic:Context'),
                models: {
                    Number: relative('../fixtures/basic/index.ts:Number'),
                },
                output: relative('./generated/basic/graphqlgen.ts')
            },
            _a['resolver-scaffolding'] = {
                output: relative('./tmp/basic/'),
                layout: 'file-per-type',
            },
            _a);
        schema = parse_1.parseSchema(config.schema);
        modelMap = parse_1.parseModels(config.models, config.output, language);
        _b = index_1.generateCode({
            schema: schema,
            language: language,
            config: config,
            modelMap: modelMap,
            prettify: true,
        }), generatedTypes = _b.generatedTypes, generatedResolvers = _b.generatedResolvers;
        expect(generatedTypes).toMatchSnapshot();
        expect(generatedResolvers).toMatchSnapshot();
        return [2 /*return*/];
    });
}); });
test('basic enum', function () { return __awaiter(_this, void 0, void 0, function () {
    var _a, language, config, schema;
    return __generator(this, function (_b) {
        language = 'typescript';
        config = (_a = {
                language: language,
                schema: relative('../fixtures/enum/schema.graphql'),
                context: relative('../fixtures/enum/types.ts:Context'),
                models: {
                    Number: relative('../fixtures/enum/types.ts:Number'),
                },
                output: relative('./generated/enum/graphqlgen.ts')
            },
            _a['resolver-scaffolding'] = {
                output: relative('./tmp/enum/'),
                layout: 'file-per-type',
            },
            _a);
        schema = parse_1.parseSchema(config.schema);
        expect(validation_1.validateConfig(config, schema)).toBe(false);
        return [2 /*return*/];
    });
}); });
test('basic union', function () { return __awaiter(_this, void 0, void 0, function () {
    var _a, language, config, schema;
    return __generator(this, function (_b) {
        language = 'typescript';
        config = (_a = {
                language: language,
                schema: relative('../fixtures/union/schema.graphql'),
                context: relative('../fixtures/union/types.ts:Context'),
                models: {
                    Number: relative('../fixtures/union/types.ts:Number'),
                },
                output: relative('./generated/union/graphqlgen.ts')
            },
            _a['resolver-scaffolding'] = {
                output: relative('./tmp/union/'),
                layout: 'file-per-type',
            },
            _a);
        schema = parse_1.parseSchema(config.schema);
        expect(validation_1.validateConfig(config, schema)).toBe(false);
        return [2 /*return*/];
    });
}); });
test('basic scalar', function () { return __awaiter(_this, void 0, void 0, function () {
    var _a, language, config, schema;
    return __generator(this, function (_b) {
        language = 'typescript';
        config = (_a = {
                language: language,
                schema: relative('../fixtures/scalar/schema.graphql'),
                context: relative('../fixtures/scalar/types.ts:Context'),
                models: {
                    Number: relative('../fixtures/scalar/types.ts:Number'),
                },
                output: relative('./generated/scalar/graphqlgen.ts')
            },
            _a['resolver-scaffolding'] = {
                output: relative('./tmp/scalar/'),
                layout: 'file-per-type',
            },
            _a);
        schema = parse_1.parseSchema(config.schema);
        expect(validation_1.validateConfig(config, schema)).toBe(false);
        return [2 /*return*/];
    });
}); });
